backend/src/org/commcare/util/FormDataUtil.java

```diff
@@ -13,55 +12,46 @@
 import org.javarosa.core.model.instance.TreeReference;
 
 /**
+ * Use the session state descriptor attached to saved forms to load case
+ * information, such as the case name.
+ *
  * @author Phillip Mates (pmates@dimagi.com)
  */
 public class FormDataUtil {
     public static String getTitleFromSession(UserSandbox userSandbox,
                                              CommCareSession session,
                                              EvaluationContext evalContext) {
         CommCareSession sessionCopy = new CommCareSession(session);
-
-        Pair<EntityDatum, String> datumEntityAndValue =
-                getDatumAndCaseId(sessionCopy, evalContext);
-        EntityDatum entityDatum = datumEntityAndValue.first;
-        String value = datumEntityAndValue.second;
-
-        if (entityDatum == null) {
-            if (value == null) {
-                return null;
-            } else {
-                return getCaseName(userSandbox, value);
-            }
-        } else {
-            return loadTitleFromEntity(entityDatum, value, evalContext, sessionCopy, userSandbox);
-        }
-    }
-
-    private static Pair<EntityDatum, String> getDatumAndCaseId(CommCareSession session,
-                                                               EvaluationContext evaluationContext) {
         String datumValue = null;
-        while (session.getFrame().getSteps().size() > 0) {
-            SessionDatum datum = session.getNeededDatum();
-            if (isCaseIdComputedDatum(datum, datumValue, session.getPoppedStep())) {
+
+        while (sessionCopy.getFrame().getSteps().size() > 0) {
+            SessionDatum datum = sessionCopy.getNeededDatum();
+            if (isCaseIdComputedDatum(datum, datumValue, sessionCopy.getPoppedStep())) {
                 // Loads case id for forms that create cases, since case id was
                 // a computed datum injected into the form. Assumes that it was
                 // the 1st computed datum... another good heuristic would be
                 // session.getPoppedStep().getId().startsWith("case_id")
-                datumValue = session.getPoppedStep().getValue();
+                datumValue = sessionCopy.getPoppedStep().getValue();
             } else if (datum instanceof EntityDatum) {
-                String tmpDatumValue = session.getPoppedStep().getValue();
+                String tmpDatumValue = sessionCopy.getPoppedStep().getValue();
                 if (tmpDatumValue != null) {
                     datumValue = tmpDatumValue;
                 }
                 if (((EntityDatum)datum).getLongDetail() == null) {
```

Comment:
can this if/else statement be re-written as just `if (((EntityDatum)datum).getLongDetail() != null) { return loadTitleFromEntity((EntityDatum)datum, datumValue, evalContext, sessionCopy, userSandbox);`, and remove the oddness of the `break`?